File: src/fable-library/System.Collections.Generic.fs
namespace System.Collections.Generic

open System

type Comparer<'T when 'T: comparison>() =
    static member Default =
        { new IComparer<'T> with
            member _.Compare(x, y) =
                LanguagePrimitives.GenericComparison x y
        }

    interface IComparer<'T> with
        member _.Compare(x, y) =
            LanguagePrimitives.GenericComparison x y

type EqualityComparer<'T when 'T: equality>() =
    static member Default =
        { new IEqualityComparer<'T> with
            member _.Equals(x, y) = LanguagePrimitives.GenericEquality x y
            member _.GetHashCode(x) = LanguagePrimitives.GenericHash x
        }

    interface IEqualityComparer<'T> with
        member _.Equals(x, y) = LanguagePrimitives.GenericEquality x y
        member _.GetHashCode(x) = LanguagePrimitives.GenericHash x

type Stack<'T> private (initialContents, initialCount) =
    let mutable contents = initialContents
    let mutable count = initialCount

    new(initialCapacity: int) =
        Stack<'T>(Array.zeroCreate<'T> (initialCapacity), 0)

    new() = Stack<'T>(4)

    new(xs: IEnumerable<'T>) =
        let arr = Array.ofSeq xs
        Stack<'T>(arr, arr.Length)

    member _.Ensure(newSize) =
        let oldSize = contents.Length

        if newSize > oldSize then
            let old = contents
            contents <- Array.zeroCreate (max newSize (oldSize * 2))
            Array.blit old 0 contents 0 count

    member _.Count = count

    member _.Pop() =
        count <- count - 1
        contents.[count]

    member _.Peek() = contents.[count - 1]

    member _.Contains(x: 'T) =
        let mutable found = false
        let mutable i = 0

        while i < count && not found do
            if Object.Equals(x, contents.[i]) then
                found <- true
            else
                i <- i + 1

        found

    member this.TryPeek(result: 'T byref) =
        if count > 0 then
            result <- this.Peek()
            true
        else
            false

    member this.TryPop(result: 'T byref) =
        if count > 0 then
            result <- this.Pop()
            true
        else
            false

    member this.Push(x) =
        this.Ensure(count + 1)
        contents.[count] <- x
        count <- count + 1

    member _.Clear() =
        count <- 0
        Array.fill contents 0 contents.Length Unchecked.defaultof<_>

    member this.TrimExcess() =
        if float count / float contents.Length > 0.9 then
            this.Ensure(count)

    member _.ToArray() =
        Array.init count (fun i -> contents.[count - 1 - i])

    interface IEnumerable<'T> with
        member _.GetEnumerator() =
            (seq {
                let mutable index = count - 1

                while index >= 0 do
                    yield contents.[index]
                    index <- index - 1
            })
                .GetEnumerator()

        member this.GetEnumerator() =
            (this :> IEnumerable<'T>).GetEnumerator() :> Collections.IEnumerator

type Queue<'T> private (initialContents, initialCount) =
    let mutable contents: 'T array = initialContents
    let mutable count = initialCount
    let mutable head = 0
    let mutable tail = initialCount

    let size () = contents.Length

    let toIndex i = (head + i) % size ()

    let ensure (requiredSize: int) =
        let newBuffer: 'T array = Array.zeroCreate requiredSize

        if head < tail then
            Array.blit contents head newBuffer 0 count
        else
            Array.blit contents head newBuffer 0 (size () - head)
            Array.blit contents 0 newBuffer (size () - head) tail

        head <- 0
        tail <- count
        contents <- newBuffer

    let toSeq () =
        seq {
            let mutable i = 0

            while i < count do
                yield contents.[i |> toIndex]
                i <- i + 1
        }

    new(initialCapacity: int) =
        if initialCapacity < 0 then
            raise (ArgumentOutOfRangeException("capacity is less than 0"))

        Queue<'T>(Array.zeroCreate<'T> (initialCapacity), 0)

    new() = Queue<'T>(4)

    new(xs: IEnumerable<'T>) =
        let arr = Array.ofSeq xs
        Queue<'T>(arr, arr.Length)

    member _.Count = count

    member _.Enqueue(value: 'T) =
        if count = size () then
            ensure (count + 1)

        contents.[tail] <- value
        tail <- (tail + 1) % size ()
        count <- count + 1

    member _.Dequeue() : 'T =
        if count = 0 then
            invalidOp "Queue is empty"

        let value = contents.[head]
        head <- (head + 1) % size ()
        count <- count - 1
        value

    member _.Peek() : 'T =
        if count = 0 then
            invalidOp "Queue is empty"

        contents.[head]

    member this.TryDequeue(result: 'T byref) : bool =
        if count = 0 then
            false
        else
            result <- this.Dequeue()
            true

    member this.TryPeek(result: 'T byref) : bool =
        if count = 0 then
            false
        else
            result <- this.Peek()
            true

    member _.Contains(x: 'T) =
        let mutable found = false
        let mutable i = 0

        while i < count && not found do
            if Object.Equals(x, contents.[i |> toIndex]) then
                found <- true
            else
                i <- i + 1

        found

    member _.Clear() =
        count <- 0
        head <- 0
        tail <- 0
        Array.fill contents 0 (size ()) Unchecked.defaultof<_>

    member _.TrimExcess() =
        if float count / float contents.Length > 0.9 then
            ensure (count)

    member _.ToArray() = toSeq () |> Seq.toArray

    member _.CopyTo(target: 'T array, start: int) =
        let mutable i = start

        for item in toSeq () do
            target.[i] <- item
            i <- i + 1

    interface IEnumerable<'T> with
        member _.GetEnumerator() = toSeq().GetEnumerator()

        member this.GetEnumerator() =
            (this :> IEnumerable<'T>).GetEnumerator() :> Collections.IEnumerator

File: src/fable-compiler-js/fable-library/System.Collections.Generic.js
import { disposeSafe, defaultOf, toIterator, getEnumerator, structuralHash, equals, compare } from "./Util.js";
import { class_type } from "./Reflection.js";
import { toArray, empty, singleton, append, enumerateWhile, delay } from "./Seq.js";
import { initialize, copyTo, fill } from "./Array.js";
import { max } from "./Double.js";
export class Comparer$1 {
    constructor() {
    }
    Compare(x, y) {
        return compare(x, y);
    }
}
export function Comparer$1_$reflection(gen0) {
    return class_type("System.Collections.Generic.Comparer`1", [gen0], Comparer$1);
}
export function Comparer$1_$ctor() {
    return new Comparer$1();
}
export function Comparer$1_get_Default() {
    return {
        Compare(x, y) {
            return compare(x, y);
        },
    };
}
export class EqualityComparer$1 {
    constructor() {
    }
    Equals(x, y) {
        return equals(x, y);
    }
    GetHashCode(x) {
        return structuralHash(x);
    }
}
export function EqualityComparer$1_$reflection(gen0) {
    return class_type("System.Collections.Generic.EqualityComparer`1", [gen0], EqualityComparer$1);
}
export function EqualityComparer$1_$ctor() {
    return new EqualityComparer$1();
}
export function EqualityComparer$1_get_Default() {
    return {
        Equals(x, y) {
            return equals(x, y);
        },
        GetHashCode(x_1) {
            return structuralHash(x_1);
        },
    };
}
export class Stack$1 {
    constructor(initialContents, initialCount) {
        this.contents = initialContents;
        this.count = (initialCount | 0);
    }
    GetEnumerator() {
        const _ = this;
        return getEnumerator(delay(() => {
            let index = _.count - 1;
            return enumerateWhile(() => (index >= 0), delay(() => append(singleton(_.contents[index]), delay(() => {
                index = ((index - 1) | 0);
                return empty();
            }))));
        }));
    }
    [Symbol.iterator]() {
        return toIterator(getEnumerator(this));
    }
    "System.Collections.IEnumerable.GetEnumerator"() {
        const this$ = this;
        return getEnumerator(this$);
    }
}
export function Stack$1_$reflection(gen0) {
    return class_type("System.Collections.Generic.Stack`1", [gen0], Stack$1);
}
function Stack$1_$ctor_Z3B4C077E(initialContents, initialCount) {
    return new Stack$1(initialContents, initialCount);
}
export function Stack$1_$ctor_Z524259A4(initialCapacity) {
    return Stack$1_$ctor_Z3B4C077E(fill(new Array(initialCapacity), 0, initialCapacity, null), 0);
}
export function Stack$1_$ctor() {
    return Stack$1_$ctor_Z524259A4(4);
}
export function Stack$1_$ctor_BB573A(xs) {
    const arr = Array.from(xs);
    return Stack$1_$ctor_Z3B4C077E(arr, arr.length);
}
export function Stack$1__Ensure_Z524259A4(_, newSize) {
    const oldSize = _.contents.length | 0;
    if (newSize > oldSize) {
        const old = _.contents;
        _.contents = fill(new Array(max(newSize, oldSize * 2)), 0, max(newSize, oldSize * 2), null);
        copyTo(old, 0, _.contents, 0, _.count);
    }
}
export function Stack$1__get_Count(_) {
    return _.count;
}
export function Stack$1__Pop(_) {
    _.count = ((_.count - 1) | 0);
    return _.contents[_.count];
}
export function Stack$1__Peek(_) {
    return _.contents[_.count - 1];
}
export function Stack$1__Contains_2B595(_, x) {
    let found = false;
    let i = 0;
    while ((i < _.count) && !found) {
        if (equals(x, _.contents[i])) {
            found = true;
        }
        else {
            i = ((i + 1) | 0);
        }
    }
    return found;
}
export function Stack$1__TryPeek_1F3DB691(this$, result) {
    if (this$.count > 0) {
        result.contents = Stack$1__Peek(this$);
        return true;
    }
    else {
        return false;
    }
}
export function Stack$1__TryPop_1F3DB691(this$, result) {
    if (this$.count > 0) {
        result.contents = Stack$1__Pop(this$);
        return true;
    }
    else {
        return false;
    }
}
export function Stack$1__Push_2B595(this$, x) {
    Stack$1__Ensure_Z524259A4(this$, this$.count + 1);
    this$.contents[this$.count] = x;
    this$.count = ((this$.count + 1) | 0);
}
export function Stack$1__Clear(_) {
    _.count = 0;
    fill(_.contents, 0, _.contents.length, defaultOf());
}
export function Stack$1__TrimExcess(this$) {
    if ((this$.count / this$.contents.length) > 0.9) {
        Stack$1__Ensure_Z524259A4(this$, this$.count);
    }
}
export function Stack$1__ToArray(_) {
    return initialize(_.count, (i) => _.contents[(_.count - 1) - i]);
}
export class Queue$1 {
    constructor(initialContents, initialCount) {
        this.contents = initialContents;
        this.count = (initialCount | 0);
        this.head = 0;
        this.tail = (initialCount | 0);
    }
    GetEnumerator() {
        const _ = this;
        return getEnumerator(Queue$1__toSeq(_));
    }
    [Symbol.iterator]() {
        return toIterator(getEnumerator(this));
    }
    "System.Collections.IEnumerable.GetEnumerator"() {
        const this$ = this;
        return getEnumerator(this$);
    }
}
export function Queue$1_$reflection(gen0) {
    return class_type("System.Collections.Generic.Queue`1", [gen0], Queue$1);
}
function Queue$1_$ctor_Z3B4C077E(initialContents, initialCount) {
    return new Queue$1(initialContents, initialCount);
}
export function Queue$1_$ctor_Z524259A4(initialCapacity) {
    if (initialCapacity < 0) {
        throw new Error("capacity is less than 0");
    }
    return Queue$1_$ctor_Z3B4C077E(fill(new Array(initialCapacity), 0, initialCapacity, null), 0);
}
export function Queue$1_$ctor() {
    return Queue$1_$ctor_Z524259A4(4);
}
export function Queue$1_$ctor_BB573A(xs) {
    const arr = Array.from(xs);
    return Queue$1_$ctor_Z3B4C077E(arr, arr.length);
}
export function Queue$1__get_Count(_) {
    return _.count;
}
export function Queue$1__Enqueue_2B595(_, value) {
    if (_.count === Queue$1__size(_)) {
        Queue$1__ensure_Z524259A4(_, _.count + 1);
    }
    _.contents[_.tail] = value;
    _.tail = (((_.tail + 1) % Queue$1__size(_)) | 0);
    _.count = ((_.count + 1) | 0);
}
export function Queue$1__Dequeue(_) {
    if (_.count === 0) {
        throw new Error("Queue is empty");
    }
    const value = _.contents[_.head];
    _.head = (((_.head + 1) % Queue$1__size(_)) | 0);
    _.count = ((_.count - 1) | 0);
    return value;
}
export function Queue$1__Peek(_) {
    if (_.count === 0) {
        throw new Error("Queue is empty");
    }
    return _.contents[_.head];
}
export function Queue$1__TryDequeue_1F3DB691(this$, result) {
    if (this$.count === 0) {
        return false;
    }
    else {
        result.contents = Queue$1__Dequeue(this$);
        return true;
    }
}
export function Queue$1__TryPeek_1F3DB691(this$, result) {
    if (this$.count === 0) {
        return false;
    }
    else {
        result.contents = Queue$1__Peek(this$);
        return true;
    }
}
export function Queue$1__Contains_2B595(_, x) {
    let found = false;
    let i = 0;
    while ((i < _.count) && !found) {
        if (equals(x, _.contents[Queue$1__toIndex_Z524259A4(_, i)])) {
            found = true;
        }
        else {
            i = ((i + 1) | 0);
        }
    }
    return found;
}
export function Queue$1__Clear(_) {
    _.count = 0;
    _.head = 0;
    _.tail = 0;
    fill(_.contents, 0, Queue$1__size(_), defaultOf());
}
export function Queue$1__TrimExcess(_) {
    if ((_.count / _.contents.length) > 0.9) {
        Queue$1__ensure_Z524259A4(_, _.count);
    }
}
export function Queue$1__ToArray(_) {
    return toArray(Queue$1__toSeq(_));
}
export function Queue$1__CopyTo_Z3B4C077E(_, target, start) {
    let i = start;
    const enumerator = getEnumerator(Queue$1__toSeq(_));
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const item = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
            target[i] = item;
            i = ((i + 1) | 0);
        }
    }
    finally {
        disposeSafe(enumerator);
    }
}
export function Queue$1__size(this$) {
    return this$.contents.length;
}
export function Queue$1__toIndex_Z524259A4(this$, i) {
    return (this$.head + i) % Queue$1__size(this$);
}
export function Queue$1__ensure_Z524259A4(this$, requiredSize) {
    const newBuffer = fill(new Array(requiredSize), 0, requiredSize, null);
    if (this$.head < this$.tail) {
        copyTo(this$.contents, this$.head, newBuffer, 0, this$.count);
    }
    else {
        copyTo(this$.contents, this$.head, newBuffer, 0, Queue$1__size(this$) - this$.head);
        copyTo(this$.contents, 0, newBuffer, Queue$1__size(this$) - this$.head, this$.tail);
    }
    this$.head = 0;
    this$.tail = (this$.count | 0);
    this$.contents = newBuffer;
}
export function Queue$1__toSeq(this$) {
    return delay(() => {
        let i = 0;
        return enumerateWhile(() => (i < this$.count), delay(() => append(singleton(this$.contents[Queue$1__toIndex_Z524259A4(this$, i)]), delay(() => {
            i = ((i + 1) | 0);
            return empty();
        }))));
    });
}
